core/tests/block.rs

```diff
@@ -423,6 +424,59 @@ fn serialize_deserialize_block_header() {
 	assert_eq!(header1, header2);
 }
 
+fn set_pow(header: &mut BlockHeader) {
+	// Set valid pow on the block as we will test deserialization of this "untrusted" from the network.
+	let edge_bits = global::min_edge_bits();
+	header.pow.proof.edge_bits = edge_bits;
+	pow::pow_size(
+		header,
+		pow::Difficulty::min(),
+		global::proofsize(),
+		edge_bits,
+	)
+	.unwrap();
+}
+
+#[test]
+fn deserialize_untrusted_header_weight() {
+	test_setup();
+	let keychain = ExtKeychain::from_random_seed(false).unwrap();
+	let builder = ProofBuilder::new(&keychain);
+	let prev = BlockHeader::default();
+	let key_id = ExtKeychain::derive_key_id(1, 1, 0, 0, 0);
+	let mut b = new_block(&[], &keychain, &builder, &prev, &key_id);
+
+	// Set excessively large output mmr size on the header.
+	b.header.output_mmr_size = 10_000;
+	b.header.kernel_mmr_size = 0;
+	set_pow(&mut b.header);
+
+	let mut vec = Vec::new();
+	ser::serialize_default(&mut vec, &b.header).expect("serialization failed");
+	let res: Result<UntrustedBlockHeader, _> = ser::deserialize_default(&mut &vec[..]);
+	assert_eq!(res.err(), Some(ser::Error::CorruptedData));
```

Comment:
so the transaction::Error::TooHeavy gets converted into a ser::Error::CorruptedData ?

Reply:
Yes. We need to return a generic `ser::Error` error and in most cases we use `CorruptedData` for low level serialization issues.

Reply:
Actually this is not happening because of `transaction::Error::TooHeavy`.
We have an additional validation check that happens during deserialization of a block header. This is where the "global weight" is used.

https://github.com/mimblewimble/grin/blob/96e8bfa34bc9718af4be99affbb7aa3df5414235/core/src/core/block.rs#L471-L479

This is a rough check that we can do very cheaply as part of deserialization.

> We can also verify the global weight at any given header height based on the overall output MMR size.
> That doesn't make sense in hindsight:-(

It's not perfect but it prevents somebody trying to broadcast a header with massively inflated MMR sizes.
And we can verify this without needing to read the previous header from the db.

We still do the per-header weight validation later to give us a more accurate validation.